Import local libs as extensions

Fr6jDJF opened the pull request:
**Basic loader for local libraries with optional looping tasks**.

- Custom libraries placed in the  _local-lib_ folder(configuration.default.ini) can be installed as in example "local_lib_pipinstal.sh" with:
venv/bin/python -m pip install -e local-lib/testlib
As it will be installed as *.egg-link in venv/../../site-packages, updating is a matter of updating the files in local-lib folder.
- list with libs to be loaded in configuration.default.ini
- lib has it's own libconfiguration.ini
- lib has it's own lang translation files *example lib holds help message(!test command only) that extend the main help string.
- No lib reload(just restart bot)
- No auto dependency install(TestLib performs one basic check on it's own for _feedparser_ example), need to figure out if/how to have those installed by the setup files

Commit:
Import local libs as extensions
- catching up

TerryGeng commented:
Hello. Thanks for contributing to our project.

I have noticed you made a lot of changes. Can you provide details about the functionalities you'd like to add? I guess what you are aiming for is adding additional commands and hooks in the main loop of the bot by adding "libraries". Did I get it right?

TerryGeng commented on local_lib_pipinstal.sh:
```diff
@@ -0,0 +1,7 @@
+
+# 202001108
+
+venv/bin/python -m pip install -e local-lib/test
```
We don't use pip to manage our project, since @azlux has his own auto-update engine. I think it would be messy if we choose to use pip to manage the lib install...

Fr6jDJF replied:
I was unaware of this (auto update disabled by default).

TerryGeng commented on mumbleBot.py:
```diff
@@ -33,13 +34,15 @@
 
 
 class MumbleBot:
-    version = 'git'
```
Please keep this line. Azlux has the routine for replacing this line with the correct version when doing a release. The 'git' stands for the development version and will be dealt with https://github.com/azlux/botamusique/blob/014086305d54a2b585839f34e8d279c384658596/mumbleBot.py#L209.

TerryGeng commented on local-lib/testlib/testlib/typical.py:
```diff
@@ -0,0 +1,92 @@
+    return inner_decorator
+
+
+class TestLib():
```
I think this Lib does something like reading RSS feed. Can you elaborate more? And if this Lib does do some thing, then we should give it a meaningful name like RSSLib.

Fr6jDJF replied:
This lib was pure for example/testing, I could bring it down as suggested being the example RSSLib

TerryGeng commented on local-lib/testlib/testlib/typical.py:
```diff
@@ -0,0 +1,92 @@
+        return rssfeeds
+
+
+def loadmod(bot):
```
load_mod.

TerryGeng commented on mumbleBot.py:
```diff
@@ -459,6 +505,23 @@ def _download(self, item):
 
+            # It's posible for libs to start tasks, they are managed/monitored over here.
+            for task in self.task_handle:
+                name = self.task_handle['method']['task_name']
```
I happen to be in the middle of considering optimizing the logic of the main loop to make it lighter. And now we have these dict lookups in the middle...

I think if you create a specific class like
```python
from enum import Enum

class TaskState(Enum):
    added = 0
    started = 1

class TaskHandle:
    def __init__(self, name, state, handle):
        self.name = name
        self.state = TaskState.added
        self.handle = handle
    ...
```

However, I'm just wondering why you put task initialization in the main loop?

Fr6jDJF replied:
Maybe It's better to just release the loop of already started thread tasks in main, this would also give better control of stopping/pause the task.
I currently placed it in the loop the get them running as late as possible, as they live quite on their own.

TerryGeng commented on mumbleBot.py:
```diff
@@ -862,6 +924,8 @@ def start_web_interface(addr, port):
     # ======================
     var.bot = MumbleBot(args)
+    # load local libraries from /local-lib
+    libimport.import_all_loclibs(var.bot)
```
import_all_local_libs

TerryGeng reviewed (changes requested):
Sorry that I don't have time to do a full review. I just briefly glanced through your changes and posted some comments.

I think we still need to do a lot of clean up and functionalities tests before I can merge it.

Actually, I'd appreciate if you could start an issue and discuss your plan with us before you get started...

@azlux and I are always on IRC channel #mumble at freenode.net and feel free to drop by and have some small discussion with us :)

Fr6jDJF commented:
> Hello. Thanks for contributing to our project.
> 
> I have noticed you made a lot of changes. Can you provide details about the functionalities you'd like to add? I guess what you are aiming for is adding additional commands and hooks in the main loop of the bot by adding "libraries". Did I get it right?
Correct, 
For me, it's non-general functions that interact with running services(for now I have  some broadcast info to chat), but it might just as well extend to more general libs supporting the discussions on Mumble, like IMDB look ups, game server status and such.

So a.t.m I manually inject my own code to interact with my running services, what means auto update must be disabled, aging the instance. functionality like this would be the solution for that.

TerryGeng commented on mumbleBot.py:
```diff
@@ -206,22 +203,80 @@ def ctrl_caught(self, signal, frame):
 
         self.exit = True
 
-    def get_version(self):
```
Can you explain this deletion?

TerryGeng commented on mumbleBot.py:
```diff
@@ -648,16 +718,13 @@ def clear(self):
     def stop(self):
         self.interrupt()
         self.is_pause = True
-        if len(var.playlist) > 0:
```
Can you explain this part?

Logics related to the loop and its state are messy and modifications, if not thoroughly tested, if prone to errors. Are you sure this will work?

TerryGeng replied:
As you can see, this part was introduced in https://github.com/azlux/botamusique/commit/fceb50e8943fb1f92537cd59b3b2ed16c14ce29b, to fix a bug. I think you need to keep the part unchanged.

TerryGeng commented:
Great! Thanks for the improvement.

I'd suggest you check out the ABC of python, and write a template ABC class for all Libs inside libimport.py (or I think you can rename libs into plugins and rename libimport.py into plugins.py to avoid confusion).

Another thing: I think you don't have to use force push in this PR. Using force push makes it hard for me to find out what is your improvement across each commit. You can just commit as usual and we will squash this PR before merging. What's your idea?

TerryGeng commented on mumbleBot.py:
```diff
@@ -459,6 +516,20 @@ def _download(self, item):
     def loop(self):
         while not self.exit and self.mumble.is_alive():
 
+            # Release started tasks if any
```
What do you think of moving this part to the place between L516 and 517, outside the while loop?

Fr6jDJF replied:
That would be better as there is no active monitoring by the bot.

TerryGeng commented on mumbleBot.py:
```diff
@@ -459,6 +516,20 @@ def _download(self, item):
+                for task in self.task_handle:
+                    self.log.debug(f'bot.loop: about to release task: {task}')
+                    try:
+                        if self.task_handle[task].state[0] == util.TaskState.started:
```
I'm kind of confused about "started", "released" and "stopped". Can you explain a little bit?

Fr6jDJF replied:
added: init
started: thread is running
released: task may start/resume it's job
paused: task waiting for released
stopped: exit/break task loop

Any renaming suggestions, or would you rather see just threads being released to do their stuff?

As for the interaction with the threads maybe [queue](https://docs.python.org/3.8/library/queue.html) is something to consider here, and feed the q object some addressed TaskHandle , 
For now I will leave it up to further discussion.

Fr6jDJF commented:
> Great! Thanks for the improvement.
> 
> I'd suggest you check out the ABC of python, and write a template ABC class for all Libs inside libimport.py 

I'm not known with that, I will check that out, thank you.
>(or I think you can rename libs into plugins and rename libimport.py into plugins.py to avoid confusion).

Ok, that would be:


                        shown with log.debug
                                 v

/local-lib/rssplugin/rssplugin/typ_rssplugin.Rss

            ^_________^
       would translate(shown by pip freeze) in to "local-rssplugin" by pip setup.cfg


> Another thing: I think you don't have to use force push in this PR. Using force push makes it hard for me to find out what is your improvement across each commit. You can just commit as usual and we will squash this PR before merging. What's your idea?

I know sorry, I'm not used to push PR, to me -f seemed the only way to get this PR updated again.

TerryGeng commented:
> I know sorry, I'm not used to push PR, to me -f seemed the only way to get this PR updated again.

Actually you can just keep pushing commits into your 'importlib' branch, and new commits will just show up here :).

Fr6jDJF committed:
wupd

-queued entries, to you can have them all on interval/single request
-a way to skip big entry list "spam" on first run !getrss (set to latest)

Fr6jDJF committed:
Update typ_rsslib.py

Fr6jDJF committed:
Update util.py

- removed TaskState

Fr6jDJF committed:
Update mumbleBot.py

- Removed TaskState
- Renamed function register_lib => register_plugin

Fr6jDJF committed:
Update interface.py

Fr6jDJF reviewed (commented):
updating

Fr6jDJF committed:
Update libimport.py

Fr6jDJF committed:
Update mumbleBot.py

- Forgot to remove variable for warning *line5 "import command" is shadowed by "command" in register_command()

Fr6jDJF committed:
added kwargs example(disabled/commented, because requires migration)

# In order to get rid of another bot as being the argument at commands, a suggestion is to migrate old commands.
# pack at mumbleBot.py.message_received / command_exc
#    self.cmd_handle[command_exc]['handle'](self, user, text, command_exc, parameter)
#                                       v
#    self.cmd_handle[command_exc]['handle'](bot=self, user=user, text=text, command_exc=command_exc, parameter=parameter)
#       - Requires modification typ_rssplugin.py commands for *args (to ignore it's bot object, since __init__ self.bot) *Example below @command('kwargs')
#       - Requires modification of all existing command.py commands for dealing with *args (to still obtain that bot object)

Fr6jDJF committed:
added kwargs example(disabled/commented, because requires migration)

# In order to get rid of another bot as being the argument at commands, a suggestion is to migrate old commands.
# pack at mumbleBot.py.message_received / command_exc
#    self.cmd_handle[command_exc]['handle'](self, user, text, command_exc, parameter)
#                                       v
#    self.cmd_handle[command_exc]['handle'](bot=self, user=user, text=text, command_exc=command_exc, parameter=parameter)
#       - Requires modification typ_rssplugin.py commands for *args (to ignore it's bot object, since __init__ self.bot) *Example below @command('kwargs')
#       - Requires modification of all existing command.py commands for dealing with *args (to still obtain that bot object)

TerryGeng commented:
Hello! Any update on this PR?

luca0N commented:
> Hello! Any update on this PR?

Looks like this user hasn't been active on GitHub since January 3, 2021... I would love to continue to work on this PR to get it merged on botamusique, though.

@TerryGeng, could we get a checklist going to see what's left for this to get merged?

I will look into this as soon as I have some time.
